File: 30.R
# Digit fifth powers

Args  <- commandArgs()
power <- ifelse(is.na(Args[6]), 5, as.numeric(Args[6]))

total_sum <- 0

max_per_digit <- 9 ^ power
max_digits <- 1
while ( max_digits * max_per_digit > as.numeric(paste(rep(9, max_digits),
                                                      collapse='')) ) {
    max_digits <- max_digits + 1
}

number <- 2
max_number <- 10 ^ max_digits
while ( number < max_number ) {
    digits <- as.integer(strsplit(as.character(number), split=NULL)[[1]])
    sum <- sum(digits ^ power)
    if ( sum == number ) {
        total_sum <- total_sum + number
    }
    number <- as.integer(number + 1)
}

cat(total_sum, fill=TRUE)

File: 30.js
#!/usr/bin/env node

// Digit fifth powers

var program = require('commander');

program
    .version('0.1.0')
    .description('Digit fifth powers')
    .option('-p, --power <int>', 'The power to raise digits by', Number, 5)
    .parse(process.argv);

var total_sum = 0;

// Work out maximum number of digits
// e.g.            5 * 59049 (295245) > 99999
// e.g.            6 * 59049 (354294) < 999999
var max_per_digit = Math.pow(9, program.power);
var max_digits = 1;
while (max_digits * max_per_digit > Math.pow(10, max_digits) - 1) {
    max_digits++;
}

var number = 2;
var max_number = Math.pow(10, max_digits);
while (number < max_number) {
    var digits = number.toString().split('');
    var sum = 0;
    for (var i = 0; i < digits.length; i++) {
        sum += Math.pow(parseInt(digits[i]), program.power);
    }
    if (sum == number) {
        total_sum += number;
    }
    number++;
}

console.log(total_sum);
